File: Data Structures & Algorithms/island-perimeter/submission-3.py
class Solution:
    def bfs(self, r, c, grid, visited):
        ROWS = len(grid)
        COLS = len(grid[0])
        queue = deque([])
        queue.append((r, c))
        visited.add((r,c))

        perimeter = 0

        while len(queue) > 0:
            cr, cc = queue.popleft()

            directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
            for direction in directions:
                dcr, dcc = cr + direction[0], cc + direction[1]

                if (dcr < 0 or dcr >= ROWS or dcc < 0 or dcc >= COLS):
                    perimeter += 1
                elif grid[dcr][dcc] == 0:
                    perimeter += 1
                elif (dcr, dcc) not in visited:
                    visited.add((dcr, dcc))
                    queue.append((dcr, dcc))
        
        return perimeter
            
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        ROWS = len(grid)
        COLS = len(grid[0])
        

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 1:
                    return self.bfs(r, c, grid, set())

File: Data Structures & Algorithms/max-area-of-island/submission-3.java
class Solution {
    public int bfs(int r, int c, boolean[][] visited, int[][] grid) {
        int ROWS = grid.length;
        int COLS = grid[0].length;
        int area = 1;
        
        Queue<int[]> queue = new LinkedList();
        queue.offer(new int[]{r, c});
        visited[r][c] = true;
        int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

        while (!queue.isEmpty()) {
            int[] gridItem = queue.poll();
            int gridItemRow = gridItem[0];
            int gridItemCol = gridItem[1];

            for (int[] direction: directions) {
                int directionRow = direction[0];
                int directionCol = direction[1];

                int curGridDirectionRow = gridItemRow + directionRow;
                int curGridDirectionCol = gridItemCol + directionCol;

                if (curGridDirectionRow < 0 || curGridDirectionRow >= ROWS || curGridDirectionCol < 0 || curGridDirectionCol >= COLS
                    || grid[curGridDirectionRow][curGridDirectionCol] == 0) {
                    continue;
                } else if (visited[curGridDirectionRow][curGridDirectionCol] == false) {
                    queue.offer(new int[]{curGridDirectionRow, curGridDirectionCol});
                    visited[curGridDirectionRow][curGridDirectionCol] = true;
                    area++;
                }
            }
        }

        return area;
    }
    public int maxAreaOfIsland(int[][] grid) {
        int ROWS = grid.length;
        int COLS = grid[0].length;

        boolean[][] visited = new boolean[ROWS][COLS];
        int maxArea = 0;

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (visited[r][c] == false && grid[r][c] == 1) {
                    maxArea = Math.max(maxArea, bfs(r, c, visited, grid));
                }
            }
        }

        return maxArea;
    }
}
